pkg/infradb/subscriberframework/actionbus/actionbus.go

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2023-2024 Intel Corporation, or its subsidiaries.
+// Copyright (c) 2024 Ericsson AB
+
+// Package actionbus holds implementation for subscribing and receiving actions
+package actionbus
+
+import (
+	"fmt"
+	"log"
+	"sync"
+)
+
+// ABus holds the ActionBus object
+var ABus = NewActionBus()
+
+// ActionBus holds the action bus info
+type ActionBus struct {
+	subscribers    map[string][]*Subscriber
+	actionHandlers map[string]ActionHandler
+	subscriberL    sync.RWMutex
+	mutex          sync.RWMutex
+}
+
+// Subscriber holds the info for each subscriber
+type Subscriber struct {
+	Name string
+	Ch   chan interface{}
+	Quit chan bool
+}
+
+// ActionHandler handles the action requests that arrive
+type ActionHandler interface {
+	HandleAction(string, *ActionData)
+}
+
+// ActionData holds the data for each action
+type ActionData struct {
+	ErrCh chan error
+}
+
+// StartSubscriber will be called by the modules to initialize and start listening for actions
+func (a *ActionBus) StartSubscriber(moduleName, actionType string, actionHandler ActionHandler) {
+	if !a.subscriberExist(actionType, moduleName) {
+		subscriber := a.Subscribe(moduleName, actionType, actionHandler)
+
+		go func() {
+			for {
+				select {
+				case action := <-subscriber.Ch:
+					log.Printf("\nSubscriber %s for %s received \n", moduleName, actionType)
+
+					handlerKey := moduleName + "." + actionType
+					if handler, ok := a.actionHandlers[handlerKey]; ok {
+						if actionData, ok := action.(*ActionData); ok {
+							handler.HandleAction(actionType, actionData)
+						} else {
+							log.Println("error: unexpected action type")
+						}
+					} else {
+						log.Println("error: no action handler found")
+					}
+				case <-subscriber.Quit:
+					close(subscriber.Ch)
+					return
+				}
+			}
+		}()
+	}
+}
+
+// NewActionBus initializes an ActionBus object
+func NewActionBus() *ActionBus {
+	return &ActionBus{
+		subscribers:    make(map[string][]*Subscriber),
+		actionHandlers: make(map[string]ActionHandler),
+	}
+}
+
+// NewActionData initializes an ActionData object
+func NewActionData() *ActionData {
+	return &ActionData{
+		ErrCh: make(chan error),
+	}
+}
+
+// Subscribe api provides registration of a subscriber to the given action
+func (a *ActionBus) Subscribe(moduleName, actionType string, actionHandler ActionHandler) *Subscriber {
+	a.subscriberL.Lock()
+	defer a.subscriberL.Unlock()
+
+	subscriber := &Subscriber{
+		Name: moduleName,
+		Ch:   make(chan interface{}),
+		Quit: make(chan bool),
+	}
+
+	a.subscribers[actionType] = append(a.subscribers[actionType], subscriber)
+	a.actionHandlers[moduleName+"."+actionType] = actionHandler
+
+	log.Printf("Subscriber %s registered for action %s\n", moduleName, actionType)
+	return subscriber
+}
+
+// GetSubscribers api is used to fetch the list of subscribers registered with given actionType
+func (a *ActionBus) GetSubscribers(actionType string) []*Subscriber {
+	a.mutex.RLock()
+	defer a.mutex.RUnlock()
```

Comment:
different mutexes are used to access `subscribers` this does not look correct

Reply:
Sorry didn't understand what you mean here. Can you please elaborate a bit ?

Reply:
you have a resource `a.subscribers` If you protect a resource with a mutex, it means that you need to lock the same mutex each time you access the resource. Here I see, you use 2 different mutexes to access the same resource `a.mutex` and `a.subscriberL`

Reply:
or since you use channels, I would think how we could avoid using mutexes at all by transferring ownership/making a deep copy/etc.

Reply:
I need to think about this

Reply:
Done

Reply:
but in `GetSubscribers` you still can use `RLock` with `subscriberL`, right?

Reply:
yes you are right I missed that. I will fix it

Reply:
Done